### ISSUE TITLE
Flipped RTTLayer

### ISSUE THREAD

**Issue by Germanunkol** [MyGUI/mygui#72]
Hi,

I believe this to be a bug. If you need more code examples, I can provide them.

I have multiple RTTLayers in the scene.
I use the new [LayerManager::createLayerAt](https://github.com/MyGUI/mygui/blob/532730d0565375539cd69ff2db76fab04aa2b6ee/MyGUIEngine/src/MyGUI_LayerManager.cpp#L194) function to create them on the fly.

The first layer (which holds content on it), is - for some reason - flipped vertically. By first layer, I mean [mLayerNodes[0]](https://github.com/MyGUI/mygui/blob/532730d0565375539cd69ff2db76fab04aa2b6ee/MyGUIEngine/src/MyGUI_LayerManager.cpp#L201).

Layers which are empty (to be filled later) don't seem to be affected by this (?).

Here's a screenshot showing the issue (Front, small texture is correct, back texture is flipped).
![screenshot05262015_202328854](https://cloud.githubusercontent.com/assets/1830593/7822545/95c42f90-03f6-11e5-9332-3c0e68aa8ce9.png)

For now, my (very ugly) workaround is to create a texture which is not shown before creating other textures.

**Comment by scrawl** [comment]
This is using the Ogre platform, correct?

Are you using the OpenGL or DirectX render system?

Could you step into OgreRTTexture::OgreRTTexture() and see if this piece of code

<pre>
			if (mTexture->getBuffer()->getRenderTarget()->requiresTextureFlipping())
			{
				mProjectMatrix[1][0] = -mProjectMatrix[1][0];
				mProjectMatrix[1][1] = -mProjectMatrix[1][1];
				mProjectMatrix[1][2] = -mProjectMatrix[1][2];
				mProjectMatrix[1][3] = -mProjectMatrix[1][3];
			}
</pre>
is getting executed?

**Comment by scrawl** [comment]
I believe I've found the issue. Try this fix: https://github.com/scrawl/mygui/commit/2a8289efc808c542f4bc07109abf3f2e31856948

The RenderManager's begin() was resetting the projection matrix set by the RTTLayer.

**Comment by Germanunkol** [comment]
Wow, that was quick. 
Yes, your fix works.

I should've asked earlier - I spent two evenings on this...
You're awesome! :) Thanks!

Will this get merged into the master branch then?

**Comment by scrawl** [comment]
I just happened to know where to look - not my first time of debugging flipped-texture bugs with Ogre, such as this one: https://ogre3d.atlassian.net/browse/OGRE-269

Yes, I'll merge it - was just waiting for your feedback first, because I haven't even tried building it on my end.

**Comment by Germanunkol** [comment]
Follow-up question:
Original issue is solved, but when I render the whole thing to two windows (I have a small debug window, since I'm rendering to the Oculus Rift and it's really annoying when you're trying to show something to a new user and can't see what they're currently seeing) then the left side of the view is flipped - on both windows.
![screenshot from 2015-05-28 19 50 22](https://cloud.githubusercontent.com/assets/1830593/7866919/e59ba75a-0572-11e5-9b0e-045a3f83f033.png)

Maybe this is something similarly easy to fix?

**Comment by scrawl** [comment]
Weird.

Try placing log statements in MyGUI::OgreRTTexture::begin(), MyGUI::OgreRenderManager::begin(), and Ogre::GLRenderSystem::_setProjectionMatrix. The log might give some insight.

Since one side is flipped and the other isn't, I assume that you're using a different RenderTexture on each side?

If so, why not reuse the same render texture on both viewports? They're meant to show the same thing, right?

**Comment by Germanunkol** [comment]
Hi,

No, they don't show the same thing, it's stereo rendering (assuming you 
mean the whole view. If you mean just the UI textures, then yes, they DO 
show the same thing, which is why I'm so confused... they're the same 
texture being rendered twice from two slightly different perspectives).

I'll try the Log thing, but might not get around to it for a few days...

**Comment by scrawl** [comment]
If you could prepare a minimal sample demonstrating the bug that I can compile on my end, I'll take a look.

**Comment by Germanunkol** [comment]
Thanks, I'll try that if I can't find the issue using the logging you mentioned. I'm pretty busy for the next few days, but I'll keep you posted. I hope to get around to it on Monday or Tuesday.

**Comment by Germanunkol** [comment]
I got this problem solved - for my case. The issue was that the second window also had two viewports which had setOverlaysEnabled set to true (the default). This way, the overlays were rendered multiple times a frame.

What I don't understand is why this flips the texture.

I've tried logging, but it doesn't help much, since I'm using precompiled Ogre libs atm. (and I don't really have the time to build Ogre from scratch), so I can't log the _setProjectionMatrix function.

So, I guess for my case, we can close this issue - as long as it's only one viewport which has setOverlaysEnabled( true ) and all others have setOverlaysEnabled( false ), it seems to be working fine.
